File: src/chain-observers/providers/manabar-provider.ts
import { TAccountName, EManabarType } from "@hiveio/wax";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { IManabarCollectorOptions, ManabarClassifier, TManabars } from "../classifiers/manabar-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export type TManabarProvided<TAccounts extends Array<TAccountName>> = {
  [K in TAccounts[number]]: TManabars;
};

export interface IManabarProviderData<TAccounts extends Array<TAccountName>> {
  manabarData: Partial<TManabarProvided<TAccounts>>;
};

export interface IManabarProviderOptions {
  manabarData: IManabarCollectorOptions[];
}

export class ManabarProvider<
  TAccounts extends Array<TAccountName> = Array<TAccountName>
> extends ProviderBase<IManabarProviderOptions, IManabarProviderData<TAccounts>> {
  public readonly manabarData = new Map<TAccountName, Set<EManabarType>>();

  public pushOptions(options: IManabarProviderOptions): void {
    for(const { account, manabarType } of options.manabarData) {
      const singleManabarData = this.manabarData.get(account);
      if(singleManabarData)
        singleManabarData.add(manabarType);
      else
        this.manabarData.set(account, new Set([manabarType]));
    }
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    const contexts: TRegisterEvaluationContext[] = [];
    for(const [account, manabarTypes] of this.manabarData)
      for(const manabarType of manabarTypes)
        contexts.push(ManabarClassifier.forOptions({ account, manabarType }));

    return contexts;
  }

  public get baseStructure(): IManabarProviderData<TAccounts> {
    return {
      manabarData: {}
    };
  }

  public async provide(data: TProviderEvaluationContext): Promise<IManabarProviderData<TAccounts>> {
    const result = this.baseStructure;

    const { manabarData } = await data.get(ManabarClassifier);
    for(const [account, manabarTypes] of this.manabarData) {
      if (result.manabarData[account] === undefined)
        result.manabarData[account] = {};

      for(const manabarType of manabarTypes) {
        if (manabarData[account] === undefined)
          break;

        result.manabarData[account][manabarType] = manabarData[account][manabarType];
      }
    }

    return result as IManabarProviderData<TAccounts>;
  }
}

File: python/workerbee/chain_observers/providers/manabar_provider.py
"""ManabarProvider — provides manabar data for accounts.

Mirrors src/chain-observers/providers/manabar-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .._ordered_set import OrderedSet
from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.manabar_classifier import ManabarClassifier
from ..enums import ManabarType
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import ManabarPayload, ManabarValue


class ManabarProvider(ProviderBase):
    def __init__(self) -> None:
        # Maps account -> set of manabar types
        self.manabar_data: dict[str, OrderedSet[ManabarType]] = {}

    def push_options(self, options: Any) -> None:
        for entry in options["manabar_data"]:
            account = entry["account"]
            manabar_type = ManabarType(entry["manabar_type"])
            if account in self.manabar_data:
                self.manabar_data[account].add(manabar_type)
            else:
                self.manabar_data[account] = OrderedSet([manabar_type])

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        contexts: list[TRegisterEvaluationContext] = []
        for account, manabar_types in self.manabar_data.items():
            for manabar_type in manabar_types:
                contexts.append(
                    ManabarClassifier.for_options(
                        {"account": account, "manabar_type": manabar_type},
                    ),
                )
        return contexts

    async def provide(self, data: DataEvaluationContext) -> ManabarPayload:
        manabar_result = await data.get(ManabarClassifier)
        source = manabar_result["manabar_data"]

        manabar_data: dict[str, dict[ManabarType, ManabarValue]] = {}
        for account, manabar_types in self.manabar_data.items():
            per_account = manabar_data.setdefault(account, {})

            for manabar_type in manabar_types:
                if source.get(account) is None:
                    break
                value = source[account].get(manabar_type)
                if value is not None:
                    per_account[manabar_type] = value

        return {"manabar_data": manabar_data}
